### PR TITLE
Add kubernetes models for API server, and yaml generation

### LINKED ISSUE #77: brupop 0.2.0 apiserver: Create k8s deployment targets for development
We'll use [`kube-rs`](https://github.com/kube-rs/kube-rs) to autogenerate kubernetes objects for deployments. This task tracks creating targets (or at least documentation) to create a cluster using the autogenerated objects.

### PR DESCRIPTION
**Pull request by cbgbt**
**Issue number:** #77 

**Description of changes:**
This change adds a build step which converts our Rust models into YAML files which can be deployed to kubernetes.

Included is a client which can be used to create and update the `BottlerocketNode` custom resources

**Testing done:**
The YAML files were applied to an EKS cluster, and then a `BottlerocketNodeClient` was used to create a `BottlerocketNode` custom resource, and update its status:

```
$ kubectl describe brn
Name:         brn-ip-192-168-66-198.us-west-2.compute.internal
Namespace:    brupop-bottlerocket-aws
Labels:       <none>
Annotations:  <none>
API Version:  brupop.bottlerocket.aws/v1
Kind:         BottlerocketNode
Metadata:
  Creation Timestamp:  2021-10-05T19:41:03Z
  Generation:          1
  Managed Fields:
    API Version:  brupop.bottlerocket.aws/v1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:ownerReferences:
          .:
          k:{"uid":"623ca9b3-bfc6-4c0f-b5d8-d8bd320a0a2f"}:
            .:
            f:apiVersion:
            f:kind:
            f:name:
            f:uid:
      f:spec:
        .:
        f:state:
        f:version:
      f:status:
        .:
        f:available_versions:
        f:current_state:
        f:current_version:
    Manager:    unknown
    Operation:  Update
    Time:       2021-10-05T19:41:03Z
  Owner References:
    API Version:     v1
    Kind:            BottlerocketNode
    Name:            ip-192-168-66-198.us-west-2.compute.internal
    UID:             623ca9b3-bfc6-4c0f-b5d8-d8bd320a0a2f
  Resource Version:  36761
  UID:               0d69e179-dbda-4058-bf4f-8e099b517a52
Spec:
  State:    WaitingForUpdate
  Version:  <nil>
Status:
  available_versions:
    1.3.0
  current_state:    WaitingForUpdate
  current_version:  1.2.1
Events:             <none>
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: models/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
```
Comment (jpculp) [inline models/Cargo.toml:1]: It's probably a good idea to add authors and license information to some of these Cargo tomls.

File: models/src/namespace.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+use k8s_openapi::api::core::v1::Namespace;
+use kube::api::ObjectMeta;
+use maplit::btreemap;
+
+/// Defines the brupop namespace
+pub fn brupop_namespace() -> Namespace {
+    Namespace {
+        metadata: ObjectMeta {
+            labels: Some(btreemap! {
+                "name".to_string() => "brupop".to_string()
```
Comment (arnaldo2792) [inline models/src/namespace.rs:11]: Nit: shouldn't this `brupop` be a constant somewhere?

File: models/src/apiserver.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            name: Some(BRUPOP_APISERVER_CLUSTER_ROLE.to_string()),
+            namespace: Some(NAMESPACE.to_string()),
+            ..Default::default()
+        },
+        rules: Some(vec![
+            PolicyRule {
+                api_groups: Some(vec![BRUPOP.to_string()]),
+                resources: Some(vec![
+                    "bottlerocketnodes".to_string(),
+                    "bottlerocketnodes/status".to_string(),
```
Comment (arnaldo2792) [inline models/src/apiserver.rs]: Nit: should these two be their own constants?

File: models/src/node.rs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        selector.node_uid,
+        source
+    ))]
+    CreateBottlerocketNode {
+        source: Box<dyn std::error::Error>,
+        selector: BottlerocketNodeSelector,
+    },
+
+    #[snafu(display(
+        "Unable to update BottlerocketNode status ({}, {}): {}",
```
Comment (arnaldo2792) [inline models/src/node.rs]: Nit: I remember folks usually ask to put error messages between `''`:

```suggestion
        "Unable to update BottlerocketNode status ({}, {}): '{}'",
```

File: models/src/node.rs [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+    async fn update_node_status(
+        &self,
+        selector: &BottlerocketNodeSelector,
+        status: &BottlerocketNodeStatus,
+    ) -> Result<(), BottlerocketNodeError> {
+        let br_node_patch = json!({
+            "apiVersion": constants::API_VERSION,
+            "kind": "BottlerocketNode",
+            "status": status
+        });
```
Comment (arnaldo2792) [inline models/src/node.rs]: Will this be at any point its own serializable struct? Probably with `serde_json`?

**Review by arnaldo2792 — APPROVED** [review]
:boot:

**Review by gthao313 — APPROVED** [review]

File: models/src/node.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+    state: BottlerocketNodeState,
+    version: Option<String>,
+}
+
+/// `BottlerocketNodeStatus` surfaces the current state of a bottlerocket node. The status is updated by the host agent,
+/// while the spec is updated by the brupop controller.
+#[derive(Clone, Debug, Default, Serialize, Deserialize, Eq, PartialEq, JsonSchema)]
+pub struct BottlerocketNodeStatus {
+    current_version: String,
+    available_versions: Vec<String>,
```
Comment (gthao313) [inline models/src/node.rs:90]: `available_versions` might be an empty Vec which means current version is the latest version, I think it's better to use `Option<Vec<String>>`

File: models/Cargo.toml (hunk 1 above)
Comment (cbgbt) [reply models/Cargo.toml:1]: These fell out of a rebase. I'll re-add them!

File: models/src/node.rs (hunk 6 above)
Comment (cbgbt) [reply models/src/node.rs:90]: I think it's only best to wrap a `Vec` in `Option` if there is a meaningful distinction between an empty vector, and a null one. In this case, both represent the absence of updates, so it's probably best to only allow a single representation of it if possible. I think not using `Option` is more clear here.

**Review by jpculp — APPROVED** [review]

**Pull request merged**